### ISSUE TITLE
Create zignatures pack for windows

### ISSUE THREAD

**Issue by Maijin** [radare/radare2#7310]
* [ ] Retrieve all the .lib from all the VS versions available and for all architectures (i.e `C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\lib\amd64\*`)
* [ ] Support ar:// to load archive files like .lib .a #6104 OR use 7z x *.lib
* [ ] Iterates over .lib content (only the *.obj) and do a `aaa ; zaF ; zos *.sig`
* [ ] Kill the duplicates over the files and potentially merge those *.sig into a single one per version of MSVC and per Architectures (to be defined)

**Comment by radare** [comment]
move this to extras

**Comment by XVilka** [comment]
Can be part of GSoC.

**Comment by securitykitten** [comment]
Hello, 

As part of some testing, I've generated all zignatures for the following directories
```
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\lib\amd64\ 
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\lib\arm\
```
I found this issue and thought that it could help.

I've uploaded them into this repo. 
https://github.com/securitykitten/r2_windows_zignatures

The zignatures are uniq'd per .lib file and I didn't consolidate them all into a single JSON file.

**Comment by Maijin** [comment]
Would you be able to create similar repository like https://github.com/radareorg/radare2-sigs where travis.yml contains the routine that autogenerate those zignature file?

**Comment by radare** [comment]
What about having 1 repo to store them all instead of just a repo with one useless script?

**Comment by radare** [comment]
And then r2pm packages for them

**Comment by Maijin** [comment]
We can also merge them sure

**Comment by Maijin** [comment]
ping @securitykitten

**Comment by xarkes** [comment]
> What about having 1 repo to store them all instead of just a repo with one useless script?

Agreed

**Comment by radare** [comment]
Yes. We said this over 9900 times and maijin have the repo already but empty

**Comment by securitykitten** [comment]
Happy to assist on this issue.  

It seems like there is conflicting ideas on the best way to solve this.  What's the best method for me to move forward and help?  

I can create a travis.yml file to generate these, but that would require uploading all the LIB's, methods to extract the .obj's etc... 

I also have access to more copies of Visual Studio (and historical copies) for different versions, builds, etc... if there is a need for bigger zignature set.  I can keep a set of zignatures for each version & build and if you'd like to pull them into a larger repo that'd work out great.

**Comment by Maijin** [comment]
@securitykitten update this repository https://github.com/radareorg/radare2-sigs with what is needed to generate the zignature, we can host all the files required on http://junk.rada.re/ and yeah if we can have all the versions that'd be great.

**Comment by securitykitten** [comment]
Sounds good.  I'm going to get started downloading/installing/extracting lib's from previous LIB's, which may take a while to generate.  I'll ping back in either this issue or in the telegram channel for further guidance once I've made more progress.

**Comment by Maijin** [comment]
The more you can automatize it the better 👍, we will eventually consolidate the `zaF` to include more stuff like types, function definition for example etc.

**Comment by securitykitten** [comment]
I've gone through the back catalog of Visual Studio releases and collected the lib files.

I have a couple small questions before I generate the zignatures for each file.

A lib contains a bunch of .obj files.  I'm unpacking each lib, generating zignatures for each obj and then combining/deduplicating the zignatures together for each lib. 

Using msvcrt.lib as an example, I have several different versions across distribution.  
```
./Microsoft Visual Studio 9.0/lib/msvcrt.lib
./Microsoft Visual Studio 9.0/lib/amd64/msvcrt.lib
./Microsoft Visual Studio 10.0/msvcrt.lib
./Microsoft Visual Studio 10.0/amd64/msvcrt.lib
./Microsoft Visual Studio 11.0/lib/msvcrt.lib
./Microsoft Visual Studio 11.0/lib/amd64/msvcrt.lib
./Microsoft Visual Studio 11.0/lib/arm/msvcrt.lib
./Microsoft Visual Studio 14.0/lib/msvcrt.lib
./Microsoft Visual Studio 14.0/lib/store/msvcrt.lib
./Microsoft Visual Studio 14.0/lib/store/amd64/msvcrt.lib
./Microsoft Visual Studio 14.0/lib/store/arm/msvcrt.lib
./Microsoft Visual Studio 14.0/lib/amd64/msvcrt.lib
./Microsoft Visual Studio 14.0/lib/arm/msvcrt.lib
./Microsoft Visual Studio 8/lib/msvcrt.lib
./Microsoft Visual Studio 8/lib/amd64/msvcrt.lib
./Microsoft Visual Studio 12.0/lib/msvcrt.lib
./Microsoft Visual Studio 12.0/lib/store/msvcrt.lib
./Microsoft Visual Studio 12.0/lib/store/amd64/msvcrt.lib
./Microsoft Visual Studio 12.0/lib/store/arm/msvcrt.lib
./Microsoft Visual Studio 12.0/lib/amd64/msvcrt.lib
./Microsoft Visual Studio 12.0/lib/arm/msvcrt.lib
```
Is the goal to have 1 main zignature covering all versions of msvcrt (different copies for each arch) or to have a zignature for each Visual Studio version for each lib?

In the final repo do you want something like 
```
VisualStudio12_ARM_msvcrt.zig
VisualStudio8_x86_msvcrt.zig
VisualStudio10_AMD64_msvcrt.zig
...etc.....
```
or just 3 final copies containing everything from all versions of VS?
```
x86_msvcrt.zig
ARM_msvcrt.zig
AMD64_msvcrt.zig
```

I appreciate the guidance.

**Comment by XVilka** [comment]
Please don't remove original binaries just yet. The biggest problem of current signatures format - they don't store function metainformation - prototype, arguments and their types: https://github.com/radare/radare2/issues/6947
Ideally, the best results of signatures search will not only find exact matches but also will improve analysis in the other places by types and variables inference.

**Comment by Maijin** [comment]
VisualStudio12_ARM_msvcrt.zig
VisualStudio8_x86_msvcrt.zig
VisualStudio10_AMD64_msvcrt.zig

This format is good^, yeah @XVilka that's why needs to ensure it's automated so then everytime we improve Zignature it will improve this automatically ^_^

**Comment by securitykitten** [comment]
I wanted to send an update to let everyone know the status.

Wrapping up work on the final script, doing some final testing now.  The final size of the libraries are roughly 650MB compressed.  When decompressed there is about 3GB of data to iterate through.  On my last run it took about 15-20 hours to process the full set of zignatures.  Each lib can have hundreds of .obj files to disassemble.

Here's the metrics from the last run
```
real	1241m28.977s
user	2144m51.030s
sys	133m21.364s
```
One small issue is travis compatibility.   The Microsoft implementation of AR is slightly different than the unix/bsd implementation, which causes breakage.  I'll explain.

It seems that travis is running on Ubuntu 14.04 and the version of 7zip (used for extracting obj's from libs) is 9.x on the builds, which is out of date. Newer versions of 7zip (I'm testing with 16.x) extract these perfectly.

I've tested pure python implementations (to avoid making a syscall) and none of the libraries cover the Windows case.  I've tried `libarchive`, `unix_ar`, `libarchive-c` & `arpy`, I've also looked at some of the ruby libraries, but they didn't work either.  It seems as if 7zip is the best current solution.

So with that said. The script that I have will extract, recurse, dedup and generate zignatures for all lib files.  I've got the last batch of processed zignatures in a zip if anyone would like to review them for testing purposes.  The zip can be found here -> https://github.com/securitykitten/r2_windows_zignatures/blob/master/visual_studio_zignatures.zip

If travis doesn't work for processing these files, I could set up a server that would cron nightly builds of the zignatures and push them to a zignature repository.  

I'll keep testing with travis and trying to build 7zip compatibility, and I'll keep everyone up to date on the the progress.

**Comment by Maijin** [comment]
You can force the installation of a recent 7z on the machine ? Can find the bin here https://www.7-zip.org/download.html

**Comment by Maijin** [comment]
I don't think having a third parties server is a good idea, if travis doesn't work we can try appveyor, but we can also setup a run on jenkins or one of our servers too

**Comment by securitykitten** [comment]
Thanks for the guidance.  I'll work on getting 7zip and travis working together and see if I can run some benchmarks.

**Comment by Maijin** [comment]
@securitykitten probably also need to see what's the blocker for the time too, may need to profile this at some point. maybe @jroimartin wants to take a look to see if there are ways to improve the autogeneration time without compromising the results. 

I will check the zip you provided.

**Comment by Maijin** [comment]
Also I would merge the zig into bigger consolidated one, no need to make one zig file per dll, just group them per version and archs like, sorry I misread when you propose your format:

- VisualStudio12_ARM.zig
- VisualStudio8_x86.zig
- VisualStudio10_AMD64.zig

**Comment by securitykitten** [comment]
Yep, making them into a single pack is easy.  I'll make that change today.  

I've also been working with multiprocessing vs single threaded to disassemble the objects and doing benchmarks to see where the best performance is.  This may require a bit of testing once I get travis working.

**Comment by pelijah** [comment]
It is impossible for now since r2 can't handle link time relocations in *.obj files.

**Comment by pelijah** [comment]
r2 doesn't preoperly handle link-time relocations. Also it has problems with applying symbol names. So some generated zignatures will be useless.
![изображение](https://user-images.githubusercontent.com/16493991/68525809-8f695500-02e6-11ea-9091-0bd99a9b07ae.png)

**Comment by XVilka** [comment]
See also these:
- https://github.com/Maktm/FLIRTDB
- https://github.com/push0ebp/sig-database
- https://github.com/push0ebp/ALLirt

**Comment by XVilka** [comment]
@pelijah please open a separate issues for these.

**Comment by pelijah** [comment]
I would prefer to send PR. One issue has already been fixed by me. Just let me some time.

**Comment by XVilka** [comment]
@pelijah is there anything left here?
